### PR TITLE
fix: 'hasSubdecks' [new backend] & refactor DeckAdapter & fix IndexOutOfBoundsException

### LINKED ISSUE #13794: [2.16beta01 Bug]: IndexOutOfBoundsException on DeckAdapter
```
java.lang.IndexOutOfBoundsException: Inconsistency detected. Invalid item position 8(offset:8).state:29 androidx.recyclerview.widget.RecyclerView{b2ffdd6 VFED.V... ......ID 0,0-1440,2673 #7f0901d1 app:id/files}, adapter:com.ichi2.anki.widgets.DeckAdapter@1000a4d, layout:androidx.recyclerview.widget.LinearLayoutManager@ae0eb86, context:com.ichi2.anki.DeckPicker@c21356d
	at androidx.recyclerview.widget.RecyclerView$Recycler.tryGetViewHolderForPositionByDeadline(RecyclerView.java:382)
	at androidx.recyclerview.widget.RecyclerView$Recycler.getViewForPosition(RecyclerView.java:2)
	at androidx.recyclerview.widget.RecyclerView$Recycler.getViewForPosition(RecyclerView.java:1)
	at androidx.recyclerview.widget.LinearLayoutManager$LayoutState.next(LinearLayoutManager.java:12)
	at androidx.recyclerview.widget.LinearLayoutManager.layoutChunk(LinearLayoutManager.java:1)
	at androidx.recyclerview.widget.LinearLayoutManager.fill(LinearLayoutManager.java:39)
	at androidx.recyclerview.widget.LinearLayoutManager.scrollBy(LinearLayoutManager.java:35)
	at androidx.recyclerview.widget.LinearLayoutManager.scrollVerticallyBy(LinearLayoutManager.java:7)
	at androidx.recyclerview.widget.RecyclerView.scrollStep(RecyclerView.java:40)
	at androidx.recyclerview.widget.RecyclerView$ViewFlinger.run(RecyclerView.java:116)
	at android.view.Choreographer$CallbackRecord.run(Choreographer.java:1301)
	at android.view.Choreographer$CallbackRecord.run(Choreographer.java:1309)
	at android.view.Choreographer.doCallbacks(Choreographer.java:923)
	at android.view.Choreographer.doFrame(Choreographer.java:847)
	at android.view.Choreographer$FrameDisplayEventReceiver.run(Choreographer.java:1283)
	at android.os.Handler.handleCallback(Handler.java:942)
	at android.os.Handler.dispatchMessage(Handler.java:99)
	at android.os.Looper.loopOnce(Looper.java:226)
	at android.os.Looper.loop(Looper.java:313)
	at android.app.ActivityThread.main(ActivityThread.java:8757)
	at java.lang.reflect.Method.invoke(Native Method)
	at com.android.internal.os.RuntimeInit$MethodAndArgsCaller.run(RuntimeInit.java:571)
	at com.android.internal.os.ZygoteInit.main(ZygoteInit.java:1067)
```

https://ankidroid.org/acra/app/1/bug/47707/report/3482ece4-f673-4242-9af7-5fdeb6368cf3
Android 13
Samsung phone

@david-allison said on Discord:
> * DeckAdapter issue (I suspect on Samsung phones): https://ankidroid.org/acra/app/1/bug/47707/report/3482ece4-f673-4242-9af7-5fdeb6368cf3
  * Likely a very simple fix, judging from StackOverflow

### PR DESCRIPTION
**Pull request by david-allison**
## Purpose / Description
Noticed while trying to reproduce another issue
* `hasSubdecks` adds padding to the `DeckAdapter`'s ViewHolder
* It was set if a single deck exists in the new backend. This shouldn't happen

## Fixes
- Fixes #13794

## Approach
* Understand the code
* Fix 'hasSubdecks'
* Realise code can be simplified further

I then saw an opportunity to perform cleanup for 70151c563e92029956fe0c8ee19de075347db599 which added a number of unnecessary `suspend` calls to the code. This handles most of the cleanup

Since `processNodes` was simplified, I then made it pure, and finally made `m[Current]DeckList` immutable lists, which should resolve #13794

## How Has This Been Tested?
Briefly, manually on an API33 emulator

Tested against Arthur's collection on an API 33 emulator

## Checklist
- [x] You have a descriptive commit message with a short title (first line, max 50 chars).
- [x] You have commented your code, particularly in hard-to-understand areas
- [x] You have performed a self-review of your own code
- [ ] UI changes: include screenshots of all affected screens (in particular showing any new or changed strings)
- [ ] UI Changes: You have tested your change using the [Google Accessibility Scanner](https://play.google.com/store/apps/details?id=com.google.android.apps.accessibility.auditor)

### CODE REVIEW

File: AnkiDroid/src/main/java/com/ichi2/anki/widgets/DeckAdapter.kt [hunk 1]
```diff
@@ -139,13 +139,14 @@ class DeckAdapter(private val layoutInflater: LayoutInflater, context: Context)
...
             mDeckList.clear()
             mCurrentDeckList.clear()
-            mRev = 0
-            mLrn = mRev
-            mNew = mLrn
-            mNumbersComputed = true
             mHasSubdecks = nodes.any { it.children.any() }
             currentDeckId = withCol { decks.current().optLong("id") }
             processNodes(nodes)
+            val topLevelNodes = nodes.filter { it.value.depth == 0 && it.value.shouldDisplayCounts() }
```
Comment (david-allison) [inline AnkiDroid/src/main/java/com/ichi2/anki/widgets/DeckAdapter.kt:148]: Note: depth == 0 should always be true. `nodes` is a list of tree nodes, not a flattened tree

**Review by mikehardy — APPROVED** [review]
all seems reasonable, thanks!

**Comment by Arthur-Milchior** [comment]
Not perfect. It misses the case Here the sundeck is the empty default deck. It exists in the model but should not be considered for UI

**Comment by david-allison** [comment]
> Not perfect. It misses the case Here the sundeck is the empty default deck. It exists in the model but should not be considered for UI

Could you elaborate on the issue. My suspicion is that this is an existing regression.

logic in `main`:

https://github.com/ankidroid/Anki-Android/blob/70151c563e92029956fe0c8ee19de075347db599/AnkiDroid/src/main/java/com/ichi2/anki/widgets/DeckAdapter.kt#L280-L284

Examples with (decks: `A`, `B` and `Default`):

1) Default is a subdeck of `A`

```
A
  Default
```

✅ padding - As it is the only subdeck, `Default` is visible, `nodes.size == 1`

----
2) Default was the only subdeck, and was 'deleted'

```
A
  [Default]
```
✅ no padding - The deck is 'deleted' and is therefore invisible, it is not provided in `nodes`, and therefore was there are no child decks

----
3) With another subdeck
```
A
  B
  Default
```

✅ padding - there are two children, so there are subdecks regardless of whether `Default` is hidden


```
No cards
```

`hasSubdecks` has no effect

**Comment by david-allison** [comment]
Updated with https://github.com/ankidroid/Anki-Android/pull/13815/commits/69c2414e852ab56be52fdd196ac453403e4a645c which should fix #13794

**Review by lukstbit — APPROVED** [review]
LGTM
This should probably close #13820 as well.
Leaving it as pending merge for now if Arthur has any extra comments.

File: AnkiDroid/src/main/java/com/ichi2/libanki/sched/TreeNode.kt [hunk 2]
```diff
@@ -24,4 +24,24 @@ data class TreeNode<T : Any>(val value: T) {
...
+        for (child in currentNode.children) {
+            yieldAll(child.flattenWithParent(parent = currentNode.value))
+        }
+    }
+
+    fun mapToParent(): Map<T, T?> = flattenWithParent().toMap()
+}
+
+// https://stackoverflow.com/a/6831626
+fun <T : Any> List<TreeNode<T>>.associateNodeWithParent(): Map<T, T?> {
```
Comment (Arthur-Milchior) [inline AnkiDroid/src/main/java/com/ichi2/libanki/sched/TreeNode.kt:43]: I'd find it more consistent if both function had the same name. Given they have the same semantic.

File: AnkiDroid/src/main/java/com/ichi2/anki/widgets/DeckAdapter.kt [hunk 3]
```diff
@@ -280,7 +284,14 @@ class DeckAdapter(private val layoutInflater: LayoutInflater, context: Context)
         indent.minimumWidth = width
     }
 
-    private suspend fun processNodes(nodes: List<TreeNode<AbstractDeckTreeNode>>) {
+    /**
+     * Returns a filtered and flattened view of [nodes]
```
Comment (Arthur-Milchior) [inline AnkiDroid/src/main/java/com/ichi2/anki/widgets/DeckAdapter.kt:286]: > Returns a flattened view of nodes, without the children of collapsed decks.


I admit I had a hard time understanding your documentation, until I read the code

**Review by Arthur-Milchior — APPROVED** [review]
My bad, I missed a changed in anki behavior.
So LGTM. Small nits. Especially, since there is a user facing change in behavior, I'd love the first commit to explicitly mention it

**Review by Arthur-Milchior — APPROVED** [review]
My bad, I missed a changed in anki behavior.
So LGTM. Small nits. Especially, since there is a user facing change in behavior, I'd love the first commit to explicitly mention it

File: AnkiDroid/src/main/java/com/ichi2/anki/widgets/DeckAdapter.kt [hunk 4]
```diff
@@ -140,7 +142,7 @@ class DeckAdapter(private val layoutInflater: LayoutInflater, context: Context)
             mLrn = mRev
             mNew = mLrn
             mNumbersComputed = true
-            mHasSubdecks = false
+            mHasSubdecks = nodes.any { it.children.any() }
```
[the remark this answers is not in the document]
Comment (dae) [reply AnkiDroid/src/main/java/com/ichi2/anki/widgets/DeckAdapter.kt:144]:
```
fn hide_default_deck(node: &mut DeckTreeNode) {
    for (idx, child) in node.children.iter().enumerate() {
        // we can hide the default if it has no children
        if child.deck_id == 1 && child.children.is_empty() {
            if child.level == 1 && node.children.len() == 1 {
                // can't remove if there are no other decks
            } else {
                // safe to remove
                node.children.remove(idx);
            }
            return;
        }
    }
}```

**Comment by david-allison** [comment]
Rebased

**Review by mikehardy — APPROVED** [review]
still LGTM, I believe all comments addressed

**Pull request merged**